core/deployment/src/main/java/io/quarkus/runner/bootstrap/AugmentActionImpl.java

```diff
@@ -0,0 +1,214 @@
+import io.quarkus.builder.BuildChainBuilder;
+import io.quarkus.builder.BuildExecutionBuilder;
+import io.quarkus.builder.BuildResult;
+import io.quarkus.builder.item.BuildItem;
+import io.quarkus.deployment.ExtensionLoader;
+import io.quarkus.deployment.QuarkusAugmentor;
+import io.quarkus.deployment.builditem.ApplicationClassNameBuildItem;
+import io.quarkus.deployment.builditem.BytecodeTransformerBuildItem;
+import io.quarkus.deployment.builditem.ConfigDescriptionBuildItem;
+import io.quarkus.deployment.builditem.GeneratedClassBuildItem;
+import io.quarkus.deployment.builditem.GeneratedResourceBuildItem;
+import io.quarkus.deployment.builditem.LaunchModeBuildItem;
+import io.quarkus.deployment.builditem.LiveReloadBuildItem;
+import io.quarkus.deployment.builditem.ShutdownContextBuildItem;
+import io.quarkus.deployment.pkg.builditem.ArtifactResultBuildItem;
+import io.quarkus.deployment.pkg.builditem.JarBuildItem;
+import io.quarkus.deployment.pkg.builditem.NativeImageBuildItem;
+import io.quarkus.runtime.LaunchMode;
+import io.quarkus.runtime.configuration.ProfileManager;
+
+/**
+ * The augmentation task that produces the application.
+ */
+public class AugmentActionImpl implements AugmentAction {
+
+    private final QuarkusBootstrap quarkusBootstrap;
+    private final CuratedApplication curatedApplication;
+    private final LaunchMode launchMode;
+    private final List<Consumer<BuildChainBuilder>> chainCustomizers;
+
+    /**
+     * A map that is shared between all re-runs of the same augment instance. This is
+     * only really relevant in dev mode, however it is present in all modes for consistency.
+     * 
+     */
+    private final Map<Class<?>, Object> reloadContext = new ConcurrentHashMap<>();
+
+    public AugmentActionImpl(CuratedApplication curatedApplication) {
+        this(curatedApplication, Collections.emptyList());
+    }
+
+    public AugmentActionImpl(CuratedApplication curatedApplication, List<Consumer<BuildChainBuilder>> chainCustomizers) {
+        this.quarkusBootstrap = curatedApplication.getQuarkusBootstrap();
+        this.curatedApplication = curatedApplication;
+        this.chainCustomizers = chainCustomizers;
+        this.launchMode = quarkusBootstrap.getMode() == QuarkusBootstrap.Mode.PROD ? LaunchMode.NORMAL
+                : quarkusBootstrap.getMode() == QuarkusBootstrap.Mode.TEST ? LaunchMode.TEST : LaunchMode.DEVELOPMENT;
+    }
+
+    @Override
+    public AugmentResult createProductionApplication() {
+        if (launchMode != LaunchMode.NORMAL) {
+            throw new IllegalStateException("Can only create a production application when using NORMAL launch mode");
+        }
+        BuildResult result = runAugment(true, Collections.emptySet(), ArtifactResultBuildItem.class);
+        JarBuildItem jarBuildItem = result.consumeOptional(JarBuildItem.class);
+        NativeImageBuildItem nativeImageBuildItem = result.consumeOptional(NativeImageBuildItem.class);
+        curatedApplication.close();
```

Comment:
I guess it'd be safer to close it in `finally` in case `runAugment` throws a runtime exception. In general though, I don't see `CuratedApplication` being closed anywhere else which looks like a major issue. Did I overlook something?
I get that depending on the mode it may have different lifecycles. But it'd be great to improve the design around it to make it more clear.
We do need to test prove it on Windows which is more sensitive to closing resources.

Reply:
Ah, sorry, I see RunningQuarkusApplicationImpl.

Reply:
There were some gradle failures on windows that were due to this (the app not being closed). The gradle tests create a temp dir and the test fails if the dir cannot be released.